assignment/scripts/3-music-collection.js

```diff
@@ -1 +1,144 @@
+console.log('Adding album: ', addToCollection('Garage, Inc.', 'Metallica', 1998, [{name: 'Whiskey in the Jar', duration: '5:04'}]));
+console.log('Adding album: ', addToCollection('Dance Fever', 'Florence and the Machine', 2022, [{name: 'Free', duration: '3:54'}]));
+console.log('Adding album: ', addToCollection('Orange Blood', 'Mt. Joy', 2022, [{name: 'Bathroom Light', duration: '3:08'}]));
+console.log('The full collection is: ', collection);
+
+console.log('\n -------------- Ending addToCollection, starting showCollection --------------\n ');
+
+// start showCollection
+function showCollection(array) {
+
+    // logging the number of albums
+    console.log(`The collection is ${array.length} albums.`);
+
+    // looping through each album, logging values with . operator
+    for (let album of array) {
+        i = 1;
+        console.log(`${album.title} by ${album.artist}, published in ${album.yearPublished}:`);
+        for (let track of album.trackList) {
+            console.log(`${i}. ${track.name}: ${track.duration}`);
+            i++;
+        }
+    }
+}
+// end showCollection
+
+// testing showCollection
+showCollection(collection);
+
+console.log(`\n -------------- Ending showCollection, starting findByArtist --------------\n `);
+
+// start findByArtist
+function findByArtist(artist) {
+
+    // init empty array
+    let results = [];
+
+    // loop over each object in the collections array
+    for (let val of collection) {
+
+        // if the artist property of the object = the input, add it to the results array
+        if (val.artist === artist) {
+            results.push(val);
+        }
+    }
+    // return the results array
+    return results;
+}
+// end findByArtist
+
+// findByArtist testing
+console.log('Testing findByArtist, expect 2 album objects in results array: ', findByArtist('Metallica'));
+console.log('Testing findByArtist, expect 1 album object in results array: ', findByArtist('Lumineers'));
+console.log('Testing findByArtist, expecting an empty array: ', findByArtist('Zach Bryan'));
+
+console.log(`\n -------------- Ending findByArtist, starting search --------------\n `);
+
+
+// #### Start of stretch goals ####
+
+// start search
+function search(criteria, trackName) {
+    // console.log(typeof(criteria), typeof(trackName));
+
+    // create empty array for results
+    let searchResults = [];
+
+    // check to see if the criteria is empty or null
+    if ((criteria === undefined && trackName === undefined)) {
+
+        // return the whole collection if it is
+        return collection;
+
+    } // end if criteria
+    
+    // trackName is true, search only using trackName
+    else if (trackName) {
+
+        // iterate over each object of the array
+        for (let val of collection) {
+
+            // iterate over each track in the object's trackList
+            for (let track of val.trackList) {
```

Comment:
Brilliant job keeping track of all of these loops. Using array methods like `filter` and `some` would also have helped here. You end up with the same number of loops under-the-hood, but since the details are taken care of for you, it is easier to follow what is going on.